### ISSUE TITLE
PLR0124: Extend `comparison-with-itself` to detect attribute, subscript, and call expressions

### ISSUE THREAD

**Issue by DMARKBORN** [astral-sh/ruff#24155]
### Summary

`PLR0124` currently only flags self-comparisons involving simple names (e.g., `a == a`). It does not detect self-comparisons with attribute access, subscripts, or call expressions. These are common copy-paste mistakes that can introduce subtle bugs.

## Example

[Playground | Ruff](https://play.ruff.rs/)

```python
# ✅ Detected by PLR0124
if a == a: ...
if left_count < left_count: ...

# ❌ Not detected by PLR0124
if self.x == self.x: ...
if slim_info.structure_id == slim_info.structure_id: ...
if a[0] == a[0]: ...
if obj.method() == obj.method(): ...
```

| Pattern | Example | Detected |
|---------|---------|:--------:|
| Simple name | `a == a` | ✅ |
| Simple name + other operators | `left_count < left_count` | ✅ |
| Attribute access | `self.x == self.x` | ❌ |
| Nested attribute | `slim_info.structure_id == slim_info.structure_id` | ❌ |
| Subscript | `a[0] == a[0]` | ❌ |
| Call | `obj.method() == obj.method()` | ❌ |

## Motivation

We encountered real bugs in production where a developer wrote something like:

```python
if slim_info.structure_id == slim_info.structure_id:
    do_something()
```

This was a copy-paste error — the developer intended to compare two different objects. The condition is always `True`, silently breaking the logic. `PLR0124` did not catch it because it only checks `Name` nodes.

## Suggested enhancement

It would be great if `PLR0124` could recursively compare AST nodes for structural equality, covering at least:

- `Attribute` — `self.x == self.x`, `a.b.c == a.b.c`
- `Subscript` — `a[0] == a[0]`, `d["key"] == d["key"]`
- `Call` — `obj.method() == obj.method()`

For `Call` expressions, while functions with side effects could theoretically return different results, in practice `f() == f()` is almost always a copy-paste mistake. Rare legitimate cases can be suppressed with `# noqa`.

**Comment by DMARKBORN** [comment]
Hey, just to be clear — I'm not a bot. I used AI to help write this up since English isn't my first language and this is my first issue here, so I wasn't sure about the format. The issue is real though, we hit it at work. Let me know if anything needs clarifying.

**Comment by ntBre** [comment]
Thanks for the suggestion! There was some brief prior discussion on this topic in https://github.com/astral-sh/ruff/issues/21201. As noted there, we'll just need to be a bit careful about side effects, but it seems worth a try.

**Comment by MichaReiser** [comment]
It might be worth researching why the upstream rule `PLR0124` chose to scope the rule as it is.

I do think we should not extend the rule to support method calls. Side effects in method calls aren't uncommon.

**Comment by DMARKBORN** [comment]
Thanks for the feedback!

Regarding why upstream Pylint scoped it to `Name` nodes — I looked at the [source](https://github.com/pylint-dev/pylint/blob/main/pylint/checkers/base/comparison_checker.py#L236). It only handles `nodes.Const` and `nodes.Name`, with no coverage for `Attribute` or `Subscript`. Not sure if that was intentional or just the initial scope.

Agreed that `Call` should be excluded due to side effects.

What do you think about extending the rule to just `Attribute` and `Subscript`? These are pure data accesses with no side-effect concerns. Are there other node types worth considering, or any concerns with these two?

**Comment by MichaReiser** [comment]
> These are pure data accesses with no side-effect concerns

Unfortunately, this isn't true in Python. Reading an attribute calls `__getattr__` and subscripts call `__getitem__`. Attributes can also be implemented by properties. All of those can execute arbitrary code (e.g. a property could increment its value on every read).

This may justify a separate rule, given that it is most likely wrong, but there's a small chance that your code does something funky, in which case the comparison is fine. In which case we could also consider including method calls. 

ESLint's [no-self-compare](https://github.com/oxc-project/oxc/blob/180dd0f8f60c383dff3325334b595e67845824a1/crates/oxc_linter/src/rules/eslint/no_self_compare.rs) rule accepts this risk for false positives. In fact, it also flags method calls, literals, and any other expression that "serialize" to the same string. 


PR adding the upstream rule https://github.com/pylint-dev/pylint/pull/2074

**Comment by MichaReiser** [comment]
Clippy also has a `eq_op` rule that captures the same problem. They also removed method calls, but they otherwise compare arbitrary left and right hand sides (https://rust-lang.github.io/rust-clippy/rust-1.94.0/index.html#eq_op)

**Comment by DMARKBORN** [comment]
Good point about `__getattr__` and `__getitem__`. Thanks for the references!

If the team agrees on following Clippy's `eq_op` approach — exclude calls but compare other expression types, we can open a PR for it.

**Comment by MichaReiser** [comment]
We'll need to think about how to avoid overlap with the existing rule and I'm interested what others think too.
